### PR TITLE
Use Ratio instead of Amount for versioning threshold

### PR DESCRIPTION
**Pull request by sydhds**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-versioning/src/versioning.rs [hunk 1]
```diff
@@ -921,12 +922,14 @@ impl MipStoreRaw {
...
-                .unwrap_or(&0) as f32;
-            let block_count_considered = self.stats.config.block_count_considered as f32;
-
-            let vote_ratio_ = 100.0 * network_version_count / block_count_considered;
-
-            let vote_ratio = Amount::const_init(vote_ratio_.round() as u64, 0);
+                .unwrap_or(&0); // as f32;
+
+            let vote_ratio = Ratio::new(100, 1);
+            vote_ratio.mul(Ratio::new(network_version_count, 1));
```
Comment (Leo-Besancon) [inline massa-versioning/src/versioning.rs:928]: I would directly use the method `from_integer` to construct these. It probably is also possible to multiply and divided by an integer directly.

File: massa-serialization/src/lib.rs [hunk 2]
```diff
@@ -392,3 +395,87 @@ impl Deserializer<bool> for BoolDeserializer {
         })(buffer)
     }
 }
+
+/// Serializer for Ratio
```
Comment (damip) [inline massa-serialization/src/lib.rs:399]: why do we need to serialize ratios ? I don't think we should be sending them through the wire

File: massa-versioning/src/versioning.rs [hunk 3]
```diff
@@ -921,12 +922,14 @@ impl MipStoreRaw {
...
                 .get(&mi.version)
-                .unwrap_or(&0) as f32;
-            let block_count_considered = self.stats.config.block_count_considered as f32;
-
-            let vote_ratio_ = 100.0 * network_version_count / block_count_considered;
-
-            let vote_ratio = Amount::const_init(vote_ratio_.round() as u64, 0);
+                .unwrap_or(&0); // as f32;
+
+            let vote_ratio = Ratio::new(100, 1);
```
Comment (damip) [inline massa-versioning/src/versioning.rs:927]: we should not be multiplying everything by 100 here, it is already a ratio

File: massa-versioning/src/versioning.rs [hunk 4]
```diff
@@ -921,12 +922,14 @@ impl MipStoreRaw {
                 .stats
                 .network_version_counters
                 .get(&mi.version)
-                .unwrap_or(&0) as f32;
-            let block_count_considered = self.stats.config.block_count_considered as f32;
-
-            let vote_ratio_ = 100.0 * network_version_count / block_count_considered;
-
-            let vote_ratio = Amount::const_init(vote_ratio_.round() as u64, 0);
+                .unwrap_or(&0); // as f32;
```
Comment (damip) [inline massa-versioning/src/versioning.rs:925]: "as f32" should be removed, floats are not deterministic

File: massa-models/src/config/constants.rs [hunk 5]
```diff
@@ -297,8 +300,6 @@ pub const MAX_LISTENERS_PER_PEER: u64 = 100;
 //
 // Constants used in versioning
 //
-/// Threshold to accept a new versioning
-pub const VERSIONING_THRESHOLD_TRANSITION_ACCEPTED: Amount = Amount::const_init(75, 0);
```
Comment (damip) [inline massa-models/src/config/constants.rs:301]: use Ratio::new_raw which is const

File: massa-versioning/src/versioning.rs (hunk 1 above)
Comment (sydhds) [reply massa-versioning/src/versioning.rs:928]: fixed

File: massa-versioning/src/versioning.rs (hunk 3 above)
Comment (sydhds) [reply massa-versioning/src/versioning.rs:927]: fixed

File: massa-versioning/src/versioning.rs (hunk 4 above)
Comment (sydhds) [reply massa-versioning/src/versioning.rs:925]: Old comment, removed

File: massa-versioning/src/versioning.rs [hunk 6]
```diff
@@ -921,12 +922,14 @@ impl MipStoreRaw {
...
+
+            let vote_ratio = Ratio::new(100, 1);
+            vote_ratio.mul(Ratio::new(network_version_count, 1));
+            vote_ratio.div(Ratio::new(
+                u64::try_from(self.stats.config.block_count_considered).unwrap_or(0),
+                1,
+            ));
```
[the remark this answers is not in the document]
Comment (sydhds) [reply massa-versioning/src/versioning.rs:932]: Fixed

File: massa-models/src/config/constants.rs (hunk 5 above)
Comment (sydhds) [reply massa-models/src/config/constants.rs:301]: Done

**Review by Leo-Besancon — APPROVED** [review]
Much better now :)

File: massa-serialization/src/lib.rs (hunk 2 above)
Comment (damip) [reply massa-serialization/src/lib.rs:399]: Let's open a followup on this

Comment (sydhds) [reply massa-serialization/src/lib.rs:399]: https://github.com/massalabs/massa/issues/4183

**Comment by damip** [comment]
@sydhds there are merge conflicts

**Pull request merged**